### ISSUE TITLE
condrestart during postinstall on minion upgrade 2017.7.1 to 2017.7.2 can kill minion

### ISSUE THREAD

**Issue by dmarkwat** [saltstack/salt#44476]
### Description of Issue/Question
I upgraded minions from 2017.7.1 to 2017.7.2 the other day and noticed that nearly half my minions weren't coming back to life.  I looked at logs and configurations for answers, but couldn't find any: logs were empty and the minion configurations had barely any changes from defaults beyond some security settings (e.g. signing messages and verifying the master signature).

I pulled open an RPM and found that the postinstall script was trying to restart the service:
```
$ rpm -qp --scripts salt-minion-2017.7.2-1.el6.noarch.rpm
postinstall scriptlet (using /bin/sh):
  if [ "$1" -ge "2" ] ; then
    /sbin/service salt-minion condrestart >/dev/null 2>&1 || :
  fi
preuninstall scriptlet (using /bin/sh):
  if [ $1 -eq 0 ] ; then
      /sbin/service salt-minion stop >/dev/null 2>&1
  fi
postuninstall scriptlet (using /bin/sh):
  if [ "$1" -ge "1" ] ; then
      /sbin/service salt-minion condrestart >/dev/null 2>&1 || :
  fi
```

This seems to contradict the documentation a bit under FAQs regarding [the proper way to restart a minion after upgrade](https://docs.saltstack.com/en/latest/faq.html#what-is-the-best-way-to-restart-a-salt-minion-daemon-using-salt-after-upgrade), especially since it killed my minions about half the time :)

Any thoughts or suggestions on what might need to be done from my end or SaltStack's?

### Setup
I used a slightly tweaked version of the upgrade SLS file [here](https://docs.saltstack.com/en/latest/faq.html#what-is-the-best-way-to-restart-a-salt-minion-daemon-using-salt-after-upgrade):
```
{%- if grains['os_family'] == 'Debian' %}

Disable starting services:
  file.managed:
    - name: /usr/sbin/policy-rc.d
    - user: root
    - group: root
    - mode: 0755
    - contents:
      - '#!/bin/sh'
      - exit 101
    # do not touch if already exists
    - replace: False
    - prereq:
      - pkg: Upgrade Salt Minion

{%- endif %}

Upgrade Salt Minion:
  pkg.installed:
    - name: salt-minion
    - version: 2017.7.2{% if grains['os_family'] == 'Debian' %}+ds-1{% elif grains['os_family'] in ['CentOS', 'RedHat'] %}-1.el{{ grains['osmajorrelease'] }}{% endif %}
    - order: last

Enable Salt Minion:
  service.enabled:
    - name: salt-minion
    - require:
      - pkg: Upgrade Salt Minion

{%- if grains['os_family'] == 'Debian' %}

Enable starting services:
  file.absent:
    - name: /usr/sbin/policy-rc.d
    - onchanges:
      - pkg: Upgrade Salt Minion

{%- endif %}

{%- if grains['os'] != 'Windows' %
Restart Salt Minion:
  cmd.run:
    - name: 'salt-call --local service.restart salt-minion'
    - bg: True
    - onchanges:
      - pkg: Upgrade Salt Minion
{%- endif %}
```

My workaround for this relied on the [`at` command's state](https://docs.saltstack.com/en/latest/ref/states/all/salt.states.at.html#salt.states.at.present).  I set it for `now +3 min` just before the pkg upgrade to give the salt minion plenty of time to try to restart and die.  This worked, but I still needed to issue a `salt-call --local service.restart salt-minion` on all the minions which is puzzling, but a discussion for a different thread possibly.

### Steps to Reproduce Issue
Install salt minion v2017.7.1 on CentOS/RHEL 6 machine.
Get some flight time; restart the salt master a few times, run some commands, etc.
Attempt to upgrade using the FAQ's template upgrade SLS file (modified for CentOS/RHEL 6 - see above).

### Versions Report
Master:
```
Salt Version:
           Salt: 2017.7.2
 
Dependency Versions:
           cffi: 1.6.0
       cherrypy: Not Installed
       dateutil: 1.5
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: 0.21.0
        libnacl: Not Installed
       M2Crypto: 0.21.1
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: 2.14
       pycrypto: 2.6.1
   pycryptodome: 3.4.3
         pygit2: 0.21.4
         Python: 2.7.5 (default, May  3 2017, 07:55:04)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4
 
System Versions:
           dist: redhat 7.4 Maipo
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-693.2.2.el7.x86_64
         system: Linux
        version: Red Hat Enterprise Linux Server 7.4 Maipo
```

Minion:
```
Salt Version:
           Salt: 2017.7.2
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.8.1
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.13 (default, Jul 12 2017, 17:32:34)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.5.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
 
System Versions:
           dist: redhat 6.9 Santiago
         locale: UTF-8
        machine: x86_64
        release: 2.6.32-696.6.3.el6.x86_64
         system: Linux
        version: Red Hat Enterprise Linux Server 6.9 Santiago
```

**Comment by garethgreenaway** [comment]
@dmarkwat To clarify, were you upgrading the salt minions using Salt itself? Having the RPMs install and then restart services is pretty standard for most packages, but would be complicated if you're installing/upgrading them via Salt.

**Comment by dmarkwat** [comment]
I was using Salt to install it by itself.  The FAQ I linked to uses an SLS file but doesn't specify whether it needs to be done in masterless mode or run from a master (or neither and it's just there as an example).  I was able to do it successfully on the previous upgrade, so building off that success I tried again for 2017.7.1 -> *.2 but obviously didn't work out.  Is there a better or more accepted way to manage Salt minion upgrades?
